cli/npm/resolution/snapshot.rs

```diff
@@ -0,0 +1,466 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use deno_core::anyhow::anyhow;
+use deno_core::anyhow::bail;
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use deno_core::futures;
+use deno_core::parking_lot::Mutex;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::lockfile::Lockfile;
+use crate::npm::cache::should_sync_download;
+use crate::npm::cache::NpmPackageCacheFolderId;
+use crate::npm::registry::NpmPackageVersionDistInfo;
+use crate::npm::registry::NpmRegistryApi;
+use crate::npm::registry::RealNpmRegistryApi;
+
+use super::NpmPackageId;
+use super::NpmPackageReq;
+use super::NpmResolutionPackage;
+use super::NpmVersionMatcher;
+
+/// Packages partitioned by if they are "copy" packages or not. Copy
+/// packages are used for peer dependencies.
```

Comment:
Could you add some more explanation what is "copy" package and what is its purpose?

Reply:
Sure, will do that tomorrow. I thouht I did somewhere in detail, but I can't find it.

Essentially "copy" packages are packages that have different peer dependency resolution than another package with the same name and resolution. In other words, a package is only a "copy" if it's a duplicate of another package for the purpose of having different peer dependency resolution. In pnpm this is done by creating package folders like `foo@1.0.0_bar@1.0.0+baz@1.0.0` (https://pnpm.io/how-peers-are-resolved), but we just do `foo@1.0.0_1`, `foo@1.0.0_2`, `foo@1.0.0_3`, etc.. because it's much shorter (reduces chance of hitting windows file path limit) and the folder can be reused with a different resolution in the future if necessary.